modules/core/src/smithy4s/schema/SchemaVisitor.scala

```diff
@@ -45,6 +46,13 @@ trait SchemaVisitor[F[_]] extends (Schema ~> F) { self =>
 
 }
 
+abstract class CachedSchemaVisitor[F[_]] extends SchemaVisitor[F] {
+  self =>
+  private val cache: MMap[ShapeId, Any] = MMap.empty
```

Comment:
don't cache by ShapeId, the key should be the schema itself. That is because several schemas can have the same ShapeId, such as Lazy/Bijection/Refinements